### DISCUSSION TITLE
Processing images to specific folder

### DISCUSSION THREAD

**Discussion by mfasold** [gethyas/hyas#412] [Q&A] [answered]
I would like to publish my hyas site by just copying the output of `npm run build` to my webspace.

During build, all the images from pages (leaf bundle) get processed by hyas/images into the root folder of `public`. As I don't like my root folder stuffed with images, how could I change the rendering target of images to a subfolder, for example `public/assets`? I am new to Hugo. :)

**Comment by h-enk** [comment]
I can't reproduce your issue /I don't see this behavior with e.g. [Doks v1](https://github.com/h-enk/doks/tree/v1):

- Assets directory (in images subfolder): https://v1--doks.netlify.app/images/bozhin-karaivanov-3HnVKfQZGTQ-unsplash_hu3d03a01dcc18bc5be0e67db3d8d209a6_4763481_768x0_resize_q90_h2_lanczos.webp
- Page resource: https://v1--doks.netlify.app/docs/cheat-sheets/shortcodes/bozhin-karaivanov-zs6GW81O538-unsplash_hu3d03a01dcc18bc5be0e67db3d8d209a6_3735066_768x0_resize_q90_h2_lanczos.webp

Deploy preview: https://v1--doks.netlify.app/

Do you have a repo you could share?

**Comment by mfasold** [comment]
Thank you for the prompt reply. I created [this clean repo](https://github.com/mfasold/testing-hyas-image-target) for easy reproducibility.

Essentially, it is a new hyas project (created today with `npm create hyas@latest`, bootstrap starter, recommended plugins). As the second commit, I just added an about page with an image. When you build the page, the images used in the rendered HTML are located in the root of public/ (and some additional images are indeed rendered to public/about).

**Reply by h-enk** [reply, marked as answer]
Thanks for your input! Please use the new [create-hyas](https://github.com/gethyas/create-hyas) (v1.0.3):

```bash
npm create hyas@latest
```

Or, in one command with options:

```bash
npm create hyas@latest my-bootstrap-project -- --template bootstrap-starter-rec
```

**Comment by mfasold** [comment]
Thank you, that fixes the problem!

Btw. I noted that this latest version contains quite some changes wrt. to the version I started with some months ago (removal of RSSlink & Site.Params.footer & lazyimg, new bs themes folder, renaming of hugo.yml). Hoping to see a stable documentation soon!

**Reply by h-enk** [reply]
> Hoping to see a stable documentation soon!

Working hard on that!

**Comment by mfasold** [comment]
Hi again! I observed a similar problem as the above: some bootstrap code is missing in the build.

Easy replicable case: Define a custom CSS class in `themes/my-bootstrap-theme/assets/scss/layouts/_pages.scss` and then style any element in the HTML with it. In the dev preview, all is fine. But when you build it, the contents of _pages.scss are not exported, so the styling is not available.

**Comment by h-enk** [comment]
👉 Most likely, you'll need to update your `safelist` or `whitelister` settings in `./config/postcss.config.js`.

Do you have a repo you could share?

I can't reproduce your issue:

I've added to `./themes/my-bootstrap-theme/assets/scss/layouts/_pages.scss`:

```scss
  .bg-yellowgreen {
    background-color: yellowgreen;
  }
```

And, to `./themes/my-bootstrap-theme/layouts/_default/baseof.html`:

```html
[..]
<body class="{{ delimit (.Scratch.Get "class") " " }} bg-yellowgreen">
[..]
```

Next ran:

```bash
pnpm build && pnpm preview
```

As expected:

![Snag_449bc62](https://github.com/gethyas/hyas/assets/3902872/0d115257-e6c8-47b3-ba3c-8d09f34675ea)

**Comment by mfasold** [comment]
Kindly take a look at [this repo](https://github.com/mfasold/mfasold-blog) . In `themes/my-bootstrap-theme/assets/scss/layouts/_pages.scss` I added a new class `.nav-subfolder` that is used in the header of all pages - but not included in the build. Even stranger: my blog content is in a column with `col-xl-8` that is also displayed correctly in the dev view (`npm run dev`) but not in the build. It seems that the `col-xl-8` CSS definition is missing (which of course was not added by myself).

**Reply by h-enk** [reply]
The PostCSS process only runs on a (production) build (so, not when running the development server). That's why you notice a difference. What the PostCSS process does is strip out unused CSS based on both styles and content files defined in `./config/postcss.config.js`.

You've added a `layouts` directory to the root of your project. So, if you add this directory to the content section of the latter file, you'll be fine! Like so:

```js
[..]
    purgecss({
      content: [
        './node_modules/@hyas/core/layouts/**/*.html',
        './themes/my-bootstrap-theme/layouts/**/*.html',
        './content/**/*.md',
        './layouts/**/*.html',
      ],
[..]
```

NB) My idea with the Hyas setup is to keep things in their (to me) logical place or separated if you like (following the default Hugo directories setup):

- a theme's or starter's core module (and integration module for that matter) would hold the `assets` and `layouts` directories
- the theme or starter itself would hold the `config`, `content`, `static`, and `themes` directories
- the `themes` directory could hold a theme directory with theme overrides (keeping the folder structure in place)

But, your setup also works fine (thank you Hugo for your flexibility)!

**Reply by mfasold** [reply]
Ah, now I understand. All unused styles are removed from the final CSS file using the purgecss library. Thank you a lot for pointing that out!

Thanks also for the high-level folder description. As I mentioned, the first Hyas version I tried did not have a themes directory. But even if it existed I probably would have put my layouts in the `./layouts` folder instead  `./themes/my-bootstrap-theme/layouts/` because a) all online Hugo tutorials do it that way and b) I usually would not expect a theme to know about my site structure (e.g. `layouts/blog`). So for Hugo newbie like me it would help to include `./layouts` into the configuration above.

**Comment by mfasold** [comment]
Sorry, but I have to come back to the issue with the images. I was so happy that the images are now in place, that I did not notice that the image are now not displayed! 

So the blog markdown content (in the repo above) includes an image `![Unboxing](ng_unboxing.jpg)`. The resulting HTML is 

```
<div class="img-wrapper"> 
 <picture> <source data-srcset="[MULTIPLE IMAGES HERE]" type="image/webp"></picture>
</div>
```
But nothing will be displayed. Did I miss anything? In the previous Hyas version, this worked out-of-the-box.

Also, it would be great to know how to use the image partial with a concrete image in the template. [This example](https://deploy-preview-256--gethyas.netlify.app/docs/guides/images/#example-2) did not work for me.

**Comment by h-enk** [comment]
In `config/_default/params.yml` you've set (line 7):

```yml
formats: [webp] # [original, webp] (default)
```

Reset to defaults (the original format is always needed) and it works.

This is how to use the picture  (or `img` or `figure`) partial:

```
{{ partial "picture" (dict
  "ctx" .
  "src" "startpage-alps.jpg"
  "class" "p-1 img-fluid lazyload"
  "alt" "The alps"
) }}
```

Documentation is in the making. Until then see [Usage Examples (Quick Start)](https://github.com/future-wd/hugo-responsive-images#usage-examples-quick-start)

**Reply by mfasold** [reply]
Perfect, thanks!

**Comment by mfasold** [comment]
I noticed that when I use images this way, they look blurry. That is, on initial page loading it will display a low resolution image, and it will not update it with a higher resolution image (I have a retina display). Interestingly, it will update and display the high resolution image when I inspect the page. It is the same for Firefox and Chrome browsers. Is this to be expected?

**Reply by h-enk** [reply]
> Is this to be expected?

No, This is what it should be like: https://v1--doks.netlify.app/docs/cheat-sheets/shortcodes/#image — hit F5 and see the effect

I don't see what you're seeing. I see the image booming into place (with a layout shift), not the blur-up image or blur-up effect. Same for images inserted via Markdown.

You could try running `npm run clean:build` before running `npm run dev`. It's a bit hard for me to see though your changes.

**Reply by mfasold** [reply]
On the linked page, the image is nice and sharp. Maybe I did not describe the problem well enough: I don't see a blur-up effect. The browser just does not load the high-res image, it seems to select a low-res picture instead. 

This is a screenshot of my blog post: 
<img width="865" alt="image" src="https://github.com/gethyas/hyas/assets/726676/764c6ab2-7fdc-4962-85ab-93a7e72b369b">

The only change I applied to params.yml is changing the widths where I reduced the selection a bit - and I just checked this, the problem persists with the default. Also after the suggested `npm run clean:build` command.

**Reply by mfasold** [reply]
Also, it looks sharp in Safari and on my mobile browser. So I see the low resolution image only in Desktop Chrome & Firefox...

**Reply by h-enk** [reply]
Note that, also the Hugo [image processing options](https://gohugo.io/content-management/image-processing/#processing-options) apply — See your [hugo.toml](https://github.com/mfasold/mfasold-blog/blob/main/config/_default/hugo.toml#L58-L63)

See in particular: [Hint](https://gohugo.io/content-management/image-processing/#hint)

I really don't have a clue why we're not seeing the blur-up effect (I'll give it another thought later).

**Reply by h-enk** [reply]
You could also try setting the `sizes` attribute (note that when you widen your browser a larger sized image is loaded)

See [Usage as a markdown render hook](https://github.com/future-wd/hugo-responsive-images#usage-as-a-markdown-render-hook)

More background: [HTMLImageElement: sizes property](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes)

**Reply by mfasold** [reply]
Thank you for the pointers. I did some tests. Firstly, the Hugo image processing options have no effect. The main issue is, that when you load the page, the two `source` tags and the `image` tag of the picture get the property `sizes="230px"` (in Firefox 225px for the same window size). This is clearly too small. I don't know how this property is set. You also have to open the Developer Tools _before_ you load the page, because when clicking Inspect afterward, the size property is (correctly) set to 832px. 

I found two ways to circumvent this issue. Firstly, as suggested by you, by setting the sizes attribute manually. That is, put the following into the front matter
```
image:
  sizes: 40em
```
The other way is to replace the `loading:  lazysizes` in `params.yml` with any of the other options `auto` or `lazy` (they have the same effect). 

Since `lazysizes` seems not useful once the sizes are provided manually, I would choose the second option. But I don't understand enough of the setup to suggest a real solution for the issue.

**Comment by mfasold** [comment]
Hi! I will close this as the main question of the image location is solved. If an explanation to the image loading happens to surface, I'd be happy to be informed. Thank you again for the quick help.
